includes/admin/class-wc-admin-duplicate-product.php

```diff
@@ -215,6 +230,42 @@ private function get_product_to_duplicate( $id ) {
 
 		return $post;
 	}
+
+	/**
+	 * Generates a unique slug for a given product. We do this so that we can override the
+	 * behavior of wp_unique_post_slug(). The normal slug generation will run single
+	 * select queries on every non-unique slug, resulting in very bad performance.
+	 *
+	 * @param WC_Product $product The product to generate a slug for.
+	 */
+	private function generate_unique_slug( $product ) {
+		global $wpdb;
+
+		// We want to remove the suffix from the slug so that we can find the maximum suffix using this root slug.
+		// This will allow us to find the next-highest suffix that is unique. While this does not support gap
+		// filling, this shouldn't matter for our use-case.
+		$root_slug = preg_replace( '/-[0-9]+$/', '', $product->get_slug() );
+
+		$results = $wpdb->get_results(
+			$wpdb->prepare( "SELECT post_name FROM $wpdb->posts WHERE post_name LIKE %s AND post_type IN ( 'product', 'product_variation' )", $root_slug . '%' )
+		);
+
+		// The slug is already unique!
+		if ( empty( $results ) ) {
+			return;
+		}
+
+		// Find the maximum suffix so we can ensure uniqueness.
+		$max_suffix = 1;
+		foreach ( $results as $result ) {
+			$suffix = intval( substr( $result->post_name, strrpos( $result->post_name, '-' ) + 1 ) );
```

Comment:
Just a suggestion as I find substr with strrpos very confusing to read, perhaps use `intval( array_pop( explode( '-', $result->post_name ) ) );` instead? Up to you to decide though.